lib/galaxy/tools/execute.py

```diff
@@ -369,22 +370,29 @@ def new_execution_slices(self):
 
     def record_success(self, execution_slice, job, outputs):
         # TODO: successful_jobs need to be inserted in the correct place...
-        self.successful_jobs.append(job)
+        self.successful_jobs.append(job.to_dict(view='collection'))  # dictify so can be detached.
         self.output_datasets.extend(outputs)
+        sa_session = self.trans.sa_session
+        # TODO: in 18.05 after SA upgrade, try bulk_save_objects or a batch insert method to create
+        # these JobToOutputDatasetCollectionAssociation, ImplicitCollectionJobsJobAssociation rows
+        # instead of adding them to the SA session and ignoring them (in this thread).
```

Comment:
I was do some testing - where I create 2000 jobs with 10 parameters using regular SA and bulk methods using sqlite. Then the same for 50 parameters. For 10 parameters - there really isn't a speed up or memory decrease using the bulk insert methods unless I store hard references to the jobs in my own array. If I do that than there start to be more serious slow downs. It makes me think all the SA references are soft and the expunge isn't really needed and in fact may slow things down a bit - and simply tracking successful_jobs above as a dict is as much benefit as we are going to get from this line of optimization for now.

Reply:
Indeed, this patch didn't solve the problem, the worker died after submitting 9,000 jobs :crying_cat_face: .
Would it be possible to recover the job mapping after the worker restart? That would workaround the issue and make restarting uWSGI easier for admins.

Reply:
> So I guess my question is are you sure you are hitting memory limits in uWSGI instead of time limits

A time limit wouldn't kill the worker, right? That's what we have in the log before this patch:
```
galaxy.tools.actions INFO 2018-06-08 22:05:30,345 [p:2747,w:1,m:0] [uWSGIWorker1Core2] Verified access to datasets for Job[unflushed,tool_id=toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] (4.077 ms)
galaxy.tools.actions INFO 2018-06-08 22:05:30,346 [p:2747,w:1,m:0] [uWSGIWorker1Core2] Setup for job Job[unflushed,tool_id=toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] complete, ready to flush (12.640 ms)
galaxy.tools.actions INFO 2018-06-08 22:05:30,959 [p:2747,w:1,m:0] [uWSGIWorker1Core2] Flushed transaction for job Job[id=1143668,tool_id=toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] (612.800 ms)
galaxy.tools.execute DEBUG 2018-06-08 22:05:30,959 [p:2747,w:1,m:0] [uWSGIWorker1Core2] Tool [toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] created job [1143668] (5614.235 ms)
Fri Jun  8 22:05:31 2018 - worker 1 (pid: 2747) is taking too much time to die...NO MERCY !!!
DAMN ! worker 1 (pid: 2747) died, killed by signal 9 :( trying respawn ...
Respawned uWSGI worker 1 (new pid: 14125)
galaxy.web.stack DEBUG 2018-06-08 22:05:32,271 [p:14125,w:1,m:0] [MainThread] Calling postfork function: <bound method UWSGIApplicationStack.set_postfork_server_name of <galaxy.web.stack.UWSGIApplicationStack object at 0x7fb85412b390>>
galaxy.web.stack DEBUG 2018-06-08 22:05:32,283 [p:14125,w:1,m:0] [MainThread] server_name set to: main.web.1
```

and after the patch (exactly the same):
```
galaxy.tools.actions INFO 2018-06-18 23:08:14,937 [p:10567,w:1,m:0] [uWSGIWorker1Core1] Verified access to datasets for Job[unflushed,tool_id=toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] (12.050 ms)
galaxy.tools.actions INFO 2018-06-18 23:08:14,947 [p:10567,w:1,m:0] [uWSGIWorker1Core1] Setup for job Job[unflushed,tool_id=toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] complete, ready to flush (36.196 ms)
galaxy.tools.actions INFO 2018-06-18 23:08:16,414 [p:10567,w:1,m:0] [uWSGIWorker1Core1] Flushed transaction for job Job[id=1170270,tool_id=toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] (1466.289 ms)
galaxy.tools.execute DEBUG 2018-06-18 23:08:16,414 [p:10567,w:1,m:0] [uWSGIWorker1Core1] Tool [toolshed.g2.bx.psu.edu/repos/earlhaminst/ete/ete_homology_classifier/3.0.0b35] created job [1170270] (4888.164 ms)
Mon Jun 18 23:08:18 2018 - worker 1 (pid: 10567) is taking too much time to die...NO MERCY !!!
DAMN ! worker 1 (pid: 10567) died, killed by signal 9 :( trying respawn ...
Respawned uWSGI worker 1 (new pid: 8070)
galaxy.web.stack DEBUG 2018-06-18 23:08:19,378 [p:8070,w:1,m:0] [MainThread] Calling postfork function: <bound method UWSGIApplicationStack.set_postfork_server_name of <galaxy.web.stack.UWSGIApplicationStack object at 0x7f0afa9d43d0>>
galaxy.web.stack DEBUG 2018-06-18 23:08:19,388 [p:8070,w:1,m:0] [MainThread] server_name set to: main.web.1
```

Reply:
> A time limit wouldn't kill the worker, right?

Oh - I was assuming it would. Let me try to find some evidence either way...

Reply:
Yeah... wasn't able to find anything either way. I definitely get that it looks like a memory issue but I'm not sure how to gel that with the observed behavior.